apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java

```diff
@@ -0,0 +1,139 @@
+    doReturn(notNullPredicate).when(builder).isNotNull(dateExpression);
+    doReturn(lowerBoundPredicate).when(builder).greaterThanOrEqualTo(dateExpression, from);
+    QueryUtils.createDateRangePredicate(builder, dateRange, dateExpression);
+    verify(builder).greaterThanOrEqualTo(dateExpression, from);
+    verify(builder).and(new Predicate[] {notNullPredicate, lowerBoundPredicate});
+  }
+
+  /**
+   * Test {@link QueryUtils#createDateRangePredicate} with only a {@link ParamPrefixEnum#LESSTHAN}
+   * upper bound {@link DateRangeParam}.
+   */
+  @Test
+  public void testCreateDateRangePredicateUpperLT() {
+    DateRangeParam dateRange =
+        new DateRangeParam(null, new DateParam(ParamPrefixEnum.LESSTHAN, UPPER_DATE));
+    LocalDate from = convertDateParamToLocalDate(dateRange.getUpperBound());
+    doReturn(notNullPredicate).when(builder).isNotNull(dateExpression);
+    doReturn(upperBoundPredicate).when(builder).lessThan(dateExpression, from);
+    QueryUtils.createDateRangePredicate(builder, dateRange, dateExpression);
+    verify(builder).lessThan(dateExpression, from);
+    verify(builder).and(new Predicate[] {notNullPredicate, upperBoundPredicate});
+  }
+
+  /**
+   * Test {@link QueryUtils#createDateRangePredicate} with only a {@link
+   * ParamPrefixEnum#LESSTHAN_OR_EQUALS} upper bound {@link DateRangeParam}.
+   */
+  @Test
+  public void testCreateDateRangePredicateUpperLE() {
+    DateRangeParam dateRange =
+        new DateRangeParam(null, new DateParam(ParamPrefixEnum.LESSTHAN_OR_EQUALS, UPPER_DATE));
+    LocalDate from = convertDateParamToLocalDate(dateRange.getUpperBound());
+    doReturn(notNullPredicate).when(builder).isNotNull(dateExpression);
+    doReturn(upperBoundPredicate).when(builder).lessThanOrEqualTo(dateExpression, from);
+    QueryUtils.createDateRangePredicate(builder, dateRange, dateExpression);
+    verify(builder).lessThanOrEqualTo(dateExpression, from);
+    verify(builder).and(new Predicate[] {notNullPredicate, upperBoundPredicate});
+  }
+
+  /**
+   * Test {@link QueryUtils#createDateRangePredicate} with both an upper and lower bound {@link
+   * DateRangeParam}.
+   */
+  @Test
+  public void testCreateDateRangePredicateBothBounds() {
+    DateRangeParam dateRange =
+        new DateRangeParam(
+            new DateParam(ParamPrefixEnum.GREATERTHAN, LOWER_DATE),
+            new DateParam(ParamPrefixEnum.LESSTHAN, UPPER_DATE));
+    LocalDate lower = convertDateParamToLocalDate(dateRange.getLowerBound());
+    LocalDate upper = convertDateParamToLocalDate(dateRange.getUpperBound());
+    doReturn(notNullPredicate).when(builder).isNotNull(dateExpression);
+    doReturn(lowerBoundPredicate).when(builder).greaterThan(dateExpression, lower);
+    doReturn(upperBoundPredicate).when(builder).lessThan(dateExpression, upper);
+    QueryUtils.createDateRangePredicate(builder, dateRange, dateExpression);
+    verify(builder).and(notNullPredicate, lowerBoundPredicate, upperBoundPredicate);
```

Comment:
The general flow I use for mockito tests is some or all of the following:
- use doReturn/doAnswer to set the return values of methods that we expect to be called.
- call the method being tested
- use ordinary assertX methods to verify the result (if applicable)
- use verify calls to ensure key methods that I expected to be called were actually called.

The `verify()` method returns a stub for the `builder` mock or spy.  That stub has all of the same methods as the the real builder.  We then call the method that we want to verify was actually called during the test and pass it the arguments we expect it to have been called with (or matchers like `eq()`, or `any()`).  In this case `and()` is the method on the builder mock that we want to verify was called.  Mockito will verify that the builder's `and()` method was called with the three parameters that we specified.

So in this case we are just verifing that the builder's `and` method was called and that it was passed our three predicates as expected.  If the conditional logic in the method had worked wrong it might have been called with different predicates or just one of them.

Reply:
Thanks for explaining! So it's a different way of doing `verify(barMock, times(1)).someMethod(param1, param2);` which is the way I am used to doing it. Just didn't recognize that specific syntax.